src/withdrawals/tasks.py

```diff
@@ -0,0 +1,351 @@
+import logging
+
+from eth_typing import ChecksumAddress, HexStr
+from sw_utils import ChainHead, InterruptHandler, ProtocolConfig, ValidatorStatus
+from web3 import Web3
+from web3.types import BlockNumber, Gwei
+
+from src.common.app_state import AppState
+from src.common.clients import consensus_client, execution_client
+from src.common.consensus import get_chain_finalized_head
+from src.common.contracts import VaultContract
+from src.common.execution import get_execution_request_fee, get_protocol_config
+from src.common.tasks import BaseTask
+from src.common.utils import calc_slot_by_block_number, format_error, round_down
+from src.config.settings import (
+    MAX_WITHDRAWAL_REQUEST_FEE,
+    MIN_ACTIVATION_BALANCE_GWEI,
+    MIN_WITHDRAWAL_AMOUNT_GWEI,
+    PARTIAL_WITHDRAWALS_INTERVAL,
+    settings,
+)
+from src.validators.consensus import fetch_consensus_validators
+from src.validators.database import VaultValidatorCrud
+from src.validators.oracles import poll_active_exits
+from src.validators.typings import ConsensusValidator
+from src.withdrawals.assets import get_queued_assets
+
+logger = logging.getLogger(__name__)
+
+
+class WithdrawalsTask(BaseTask):
+    async def process_block(self, interrupt_handler: InterruptHandler) -> None:
+        """
+        Every N hours check the exit queue and submit partial withdrawals if needed.
+        """
+        chain_head = await get_chain_finalized_head()
+        protocol_config = await get_protocol_config()
+        for vault_address in settings.vaults:
+            await self.process_withdrawals(
+                vault_address=vault_address,
+                chain_head=chain_head,
+                protocol_config=protocol_config,
+            )
+
+    async def process_withdrawals(
+        self,
+        vault_address: ChecksumAddress,
+        chain_head: ChainHead,
+        protocol_config: ProtocolConfig,
+    ) -> None:
+        app_state = AppState()
+        if not await self._check_withdrawals_block(
+            app_state, vault_address, chain_head.block_number
+        ):
+            return
+
+        vault_validators = VaultValidatorCrud().get_vault_validators(
+            vault_address=vault_address,
+        )
+        consensus_validators = await fetch_consensus_validators(
+            [val.public_key for val in vault_validators]
+        )
+        oracle_exiting_validators = await _fetch_oracle_exiting_validators(
+            consensus_validators, protocol_config
+        )
+
+        queued_assets = await get_queued_assets(
+            vault_address=vault_address,
+            consensus_validators=consensus_validators,
+            oracle_exiting_validators=oracle_exiting_validators,
+            chain_head=chain_head,
+        )
+        if queued_assets < MIN_WITHDRAWAL_AMOUNT_GWEI:
+            app_state.partial_withdrawal_cache[vault_address] = chain_head.block_number
+            return
+
+        current_fee = await get_execution_request_fee(
```

Comment:
Would be great to check whether the withdrawals queue is not full. In consensus specs (https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md):
```python
    # If partial withdrawal queue is full, only full exits are processed
    if (
        len(state.pending_partial_withdrawals) == PENDING_PARTIAL_WITHDRAWALS_LIMIT
        and not is_full_exit_request
    ):
        return
```